From e43b777114fd586e1f7e0635d0491b286b178078 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Fri, 28 Jul 2017 18:20:42 +0800
Subject: [PATCH 2315/5242] MLK-16013-37 staging: typec: Power swap force
 discharge for source

commit  60376f48b4ab5a8c48070161095895ae9754c545 from
https://source.codeaurora.org/external/imx/linux-imx.git

VBus off only means the vbus falls to be below 4v, we can use vbus
force discharge and vbus low alarm to go forward.

Reviewed-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/typec/tcpm.c |   16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/typec/tcpm.c b/drivers/usb/typec/tcpm.c
index 4a32b03..34d1cab 100644
--- a/drivers/usb/typec/tcpm.c
+++ b/drivers/usb/typec/tcpm.c
@@ -3816,7 +3816,13 @@ static void _tcpm_pd_vbus_off(struct tcpm_port *port)
 		break;
 
 	case PR_SWAP_SRC_SNK_TRANSITION_OFF:
-		tcpm_set_state(port, PR_SWAP_SRC_SNK_SOURCE_OFF, 0);
+		/*
+		 * At this moment, vbus may only fall to be below 4v,
+		 * we need wait tPSSourceOff and let vbus discharge
+		 * finished.
+		 */
+		if (!port->tcpc->vbus_discharge)
+			tcpm_set_state(port, PR_SWAP_SRC_SNK_SOURCE_OFF, 0);
 		break;
 
 	case PR_SWAP_SNK_SRC_SINK_OFF:
@@ -3855,8 +3861,16 @@ static void _tcpm_vbus_discharge(struct tcpm_port *port, bool on)
 {
 	tcpm_log_force(port, "%s force discharge", on ? "Enable":"Disable");
 
+	/*
+	 * By vbus discharge and low alarm, now we can disable
+	 * vbus discharge.
+	 */
 	if (port->tcpc && port->tcpc->vbus_discharge)
 		port->tcpc->vbus_discharge(port->tcpc, false);
+
+	/* We can transit to PR_SWAP_SRC_SNK_SOURCE_OFF safely */
+	if (port->state == PR_SWAP_SRC_SNK_TRANSITION_OFF)
+		tcpm_set_state(port, PR_SWAP_SRC_SNK_SOURCE_OFF, 0);
 }
 
 static void tcpm_pd_event_handler(struct work_struct *work)
-- 
1.7.9.5

